Add support for the label listing

lachmanfrantisek opened the pull request:
- Add listing labels with ability to copy them to other repo.


```
Usage: tool labels [OPTIONS] [REPO]

  List the labels for the project. This is how you can select a repository
  for Github: <namespace>/<project>.

Options:
  -s, --service TEXT  Name of the git service (e.g. github/gitlab).
  -c, --copy TEXT     Copy labels to another repo.
  --help              Show this message and exit.
```


```
$ tool labels -c user-cont/colin -c user-cont/conu
╒══════════════════════════════════╤════════╤══╕
│ bug                              │ d73a4a │  │
├──────────────────────────────────┼────────┼──┤
│ duplicate                        │ cfd3d7 │  │
├──────────────────────────────────┼────────┼──┤
│ enhancement                      │ a2eeef │  │
├──────────────────────────────────┼────────┼──┤
│ good first issue                 │ 7057ff │  │
├──────────────────────────────────┼────────┼──┤
│ help wanted                      │ 008672 │  │
├──────────────────────────────────┼────────┼──┤
│ invalid                          │ e4e669 │  │
├──────────────────────────────────┼────────┼──┤
│ label-for-testing-the-label-sync │ 0e8a16 │  │
├──────────────────────────────────┼────────┼──┤
│ question                         │ d876e3 │  │
├──────────────────────────────────┼────────┼──┤
│ wontfix                          │ ffffff │  │
╘══════════════════════════════════╧════════╧══╛
1 labels of 9 moved to user-cont/colin
1 labels of 9 moved to user-cont/conu
```

lachmanfrantisek commented:
There is only a problem with the descriptions -- the GitHub API does return them...

TomasTomecek commented on tool/cli.py:
```diff
@@ -100,11 +100,52 @@ def list_branches():
     print(tabulate(a.list_branches(), tablefmt="fancy_grid"))
 
 
+@click.command(name="labels",
```
this could be confusing to new users: could we split the command to two? `list-labels` and `sync` or `update` `-labels`

TomasTomecek commented on tool/cli.py:
```diff
@@ -100,11 +100,52 @@ def list_branches():
+                    "This is how you can select a repository for Github: <namespace>/<project>.")
+@click.option('--service', "-s", type=click.STRING, default="github",
+              help="Name of the git service (e.g. github/gitlab).")
+@click.option('--copy', '-c', type=click.STRING, required=False, multiple=True,
```
see above, let's move this functionality to a new command

TomasTomecek commented on tool/cli.py:
```diff
@@ -100,11 +100,52 @@ def list_branches():
+        (
+            label.name,
+            label.color,
+            label.description
```
is it that the library doesn't support this or the api itself?

lachmanfrantisek replied:
It supports, but the result actually does not contain that...
https://api.github.com/repos/user-cont/tool/labels

TomasTomecek replied:
lol, hopefully ~~github~~ microsoft fixes that sooner or later

TomasTomecek commented on tool/services/github_service.py:
```diff
@@ -118,4 +118,34 @@ def create_pull_request(self, target_remote, target_branch, current_branch):
+        """
+        return list(self.repo.get_labels())
+
+    def update_labels(self, labels):
```
what do we do with labels which are in current repo and not in `labels`?

lachmanfrantisek replied:
That's a question. Some `--force` option?

Some labels can differ only a little (`good first issue` x `good-first-issue`) and contain some issues already...

TomasTomecek replied:
okay, that makes sense, let's not implement the purge functionality yet then

lachmanfrantisek commented:
Subcommand split:

```
$ tool list_labels --help
Usage: tool list_labels [OPTIONS] [REPO]

  List the labels for the project. This is how you can select a repository
  for Github: <namespace>/<project>.

Options:
  -s, --service TEXT  Name of the git service (e.g. github/gitlab).
  --help              Show this message and exit.
```

```
$ tool update_labels --help
Usage: tool update_labels [OPTIONS] DESTINATION [SOURCE]

  Update labels of other project. Multiple destinations can be set by
  joining them with semicolon. This is how you can select a repository for
  Github: <namespace>/<project>.

Options:
  -s, --service TEXT  Name of the git service (e.g. github/gitlab).
  --help              Show this message and exit.
```
-----

@TomasTomecek I have two questions:

- Any idea to achieve multiple destinations than this:
```
$ tool update_labels user-cont/colin\;user-cont/conu user-cont/tool
```
- Do we need to be consistent with the `REPO` argument in other commands? Is the `destination` ok?

Thanks.

lachmanfrantisek commented:
I've also added support for the gitlab. Some notes about that:

- The colour format differ (Github does not have `#` at the start). --> Currently, I am converting the colours on the update. Does it make sense to use some common format for `list-labels`?
- I've added `--source-service` to allow different service for the source and destination. (If there are multiple destinations, all will have the same service.)
- In the future, we can have some file import/export for the sync.

TomasTomecek commented:
Any idea to achieve multiple destinations than this:

    $ tool update-labels user-cont/colin\;user-cont/conu user-cont/tool

how about `$ tool update-labels --source-repo https://github.com/user-cont/kwaciaren user-cont/x user-cont/y`. `--source-repo` would default to repo in PWD.

    Do we need to be consistent with the REPO argument in other commands? Is the destination ok?

Yes, that's fine. I'm working on configuration PR and let's discuss over there once I open it.

lachmanfrantisek commented:
@TomasTomecek I've done as you wrote...

TomasTomecek commented:
Well done!
